### PR TITLE
Fix CUDA incompatibility issues with Eigen and GCC

### PR DESCRIPTION
**Pull request by shrijitsingh99**
As discussed in #4101

### CODE REVIEW

File: .dev/docker/env/Dockerfile [hunk 1]
```diff
@@ -2,15 +2,15 @@
...
+ARG CUDA_VERSION="9.2"
 ARG UBUNTU_DISTRO="16.04"
-# Known conflicts:
-# * 9.1 is an issue with 16.04 for Eigen
 ARG BASE_CUDA_IMAGE=${USE_CUDA:+"nvidia/cuda:${CUDA_VERSION}-devel-ubuntu${UBUNTU_DISTRO}"}
 ARG BASE_IMAGE=${BASE_CUDA_IMAGE:-"ubuntu:${UBUNTU_DISTRO}"}
 
 FROM ${BASE_IMAGE}
 
+ARG USE_CUDA
```
Comment (kunaltyagi) [inline .dev/docker/env/Dockerfile:12]: Why did you repeat L4 here?

Comment (shrijitsingh99) [reply .dev/docker/env/Dockerfile:12]: Not needed now, earlier I was checking if the flag was set to then install Eigen 3.3.7

**Comment by kunaltyagi** [comment]
What's the GCC version in 18.04? I'm asking because it can happen that we accidentally stumble on a bug in older GCC if we use CI with GCC-7 in 16.04 till it EOLs

**Comment by shrijitsingh99** [comment]
> What's the GCC version in 18.04? I'm asking because it can happen that we accidentally stumble on a bug in older GCC if we use CI with GCC-7 in 16.04 till it EOLs

Its GCC 7. I guess then it's redundant to install GCC 7 since we are not running the CI on 16.04.
Sidenote: We should definitely list the requirements of GPU module somewhere so that people don't get confused.

**Comment by kunaltyagi** [comment]
> list the requirements of GPU module somewhere

Once the CI is done, we'll know them 😆. Feel free to create an issue in the meantime

PS: mention that the requirements aren't for Windows + CUDA or MacOS + CUDA

**Comment by kunaltyagi** [comment]
In that case, I'd propose removing CUDA from 16.04 CI (and docker CI), and only supporting for 18.04 (Nvidia itself didn't support 20.04 a few days ago)

File: .dev/docker/env/Dockerfile [hunk 2]
```diff
@@ -36,8 +35,31 @@ RUN apt-get update \
       libvtk${VTK_VERSION}-dev \
       libvtk${VTK_VERSION}-qt-dev \
       qtbase5-dev \
+      software-properties-common \
  && rm -rf /var/lib/apt/lists/*
 
+# Eigen patch (https://eigen.tuxfamily.org/bz/show_bug.cgi?id=1462) to fix issue metioned
+# in https://github.com/PointCloudLibrary/pcl/issues/3729 is available in Eigen 3.3.7
+# Not needed from 20.04 since it is the default version from apt
+RUN wget -qO- https://gitlab.com/libeigen/eigen/-/archive/3.3.7/eigen-3.3.7.tar.gz | tar xz \
```
Comment (kunaltyagi) [inline .dev/docker/env/Dockerfile:44]: Install optionally using:
```sh
[ `apt show libeigen3-dev 2>/dev/null | grep '^version' -i | awk -F'[ -.]' '{print $4}'` -lt 7 ] && echo "Need to install 3.3.7"
```

**Comment by shrijitsingh99** [comment]
> In that case, I'd propose removing CUDA from 16.04 CI (and docker CI), and only supporting for 18.04 (Nvidia itself didn't support 20.04 a few days ago)
By removing CUDA do you mean removing it from the env docker image for Ubuntu 16.04?

**Comment by shrijitsingh99** [comment]
> By removing CUDA do you mean removing it from the env docker image for Ubuntu 16.04?
I Will put in separate PR for that. 
Though do we want to do that, what if someone uses the wants to use the PCL 16.04 docker image with CUDA (Very unlikely scenario though)?

**Comment by kunaltyagi** [comment]
Main aim is to have a docker image that's the "recommended and supported" setup. Makes sense to remove it from docker image too

**Comment by kunaltyagi** [comment]
MacOS test is failing... weird

File: .dev/docker/env/Dockerfile [hunk 3]
```diff
@@ -36,8 +35,32 @@ RUN apt-get update \
...
       libvtk${VTK_VERSION}-qt-dev \
       qtbase5-dev \
+      software-properties-common \
  && rm -rf /var/lib/apt/lists/*
 
+# Eigen patch (https://eigen.tuxfamily.org/bz/show_bug.cgi?id=1462) to fix issue metioned
+# in https://github.com/PointCloudLibrary/pcl/issues/3729 is available in Eigen 3.3.7
+# Not needed from 20.04 since it is the default version from apt
+RUN if [ `apt show libeigen3-dev 2>/dev/null | grep '^version' -i | awk -F'[ -.]' '{print $4}'` -lt 7 ]; then \
+ && wget -qO- https://gitlab.com/libeigen/eigen/-/archive/3.3.7/eigen-3.3.7.tar.gz | tar xz \
```
Comment (kunaltyagi) [inline .dev/docker/env/Dockerfile:45]:
```suggestion
  wget -qO- https://gitlab.com/libeigen/eigen/-/archive/3.3.7/eigen-3.3.7.tar.gz | tar xz \
```

Also, the whitespace is not uniform

**Comment by shrijitsingh99** [comment]
> Works, but I'm thinking if I've over-complicated the eigen check. Would it be better to check for the OS?

I don't think it will make a difference, will it?
Maybe we can simplify the condition to `[ pkg-config --modversion eigen3 | cut -d. -f3 -lt  7 ]`

File: .dev/docker/env/Dockerfile [hunk 4]
```diff
@@ -2,10 +2,8 @@
 # https://gitlab.com/nvidia/container-images/cuda/tree/master/dist
 # To enable cuda, use "--build-arg USE_CUDA=true" during image build process
 ARG USE_CUDA
-ARG CUDA_VERSION="9.0"
+ARG CUDA_VERSION="9.2"
```
Comment (kunaltyagi) [inline .dev/docker/env/Dockerfile:5]: Any benefits to changing this?

Comment (SergioRAgostinho) [reply .dev/docker/env/Dockerfile:5]: Maybe not although I believe I'm the one who suggested the change on Discord. I recall there were some compilation issues between 9.0 and 9.2 and I would assume people would pick the highest minor version when picking CUDA 9. That's also the only one in the 9.X releases that PyTorch is supporting.

**Pull request merged**